File: Structural_Patterns/Adapter/python/object_adapter.py
# File name     : object_adapter.py
# Description   :
# Author        :
# Date          :
#
#


class Target:
    def request(self) -> str:
        return "Target: The default target's behaviour"


class Adaptee:
    def specificRequest(self):
        return ".eetpadA eht fo ruoivaheb laicepS"


class Adapter(Target):
    def __init__(self, adaptee : Adaptee) -> None:
        self.adaptee = adaptee

    def request(self) -> str:
        return f"Adapter: (TRANSLATED) {self.adaptee.specificRequest()[::-1]}"

def clientExecute(target : Target):
    print(target.request(), end="")



if __name__ == "__main__":
    print("Client: I can work just fine with the Target objects:")

    target = Target()
    clientExecute(target)
    print("\n")

    adaptee = Adaptee()
    print("Client: The Adaptee class has a weird interface. See, I don't understand it:")
    print(f"Adaptee: {adaptee.specificRequest()}", end="\n\n")

    print("Client: But I can work with it via the Adapter:")
    adapter = Adapter(adaptee)
    clientExecute(adapter)

File: Structural_Patterns/Adapter/cpp/object_adapter/object_adapter.cpp
/*!
 *  \file       object_adapter.cpp
 *  \brief
 *
 */



#include "object_adapter.hpp"

#include <iostream>

namespace object_adapter
{
std::string Target::request() const
{
    return std::string("Target: The default target's behaviour.");
}


std::string Adaptee::specificRequest() const
{
    return std::string(".eetpadA eht fo ruoivaheb cificepS");
}


Adapter::Adapter(Adaptee *adaptee) : m_adaptee(adaptee)
{
}


std::string Adapter::request() const
{
    std::string to_reverse = m_adaptee->specificRequest();
    std::reverse(to_reverse.begin(), to_reverse.end());
    
    return std::string("Adapter: (TRANSLATED) " + to_reverse);
}


/*!
 *  \fn     execute
 *  \brief  The client code supports all classes that follow the Target's interface.
 */
void Client::execute(const Target *target)
{
    std::cout << target->request() << std::endl;
}
}
